token/program-2022/src/extension/group_member_pointer/processor.rs

```diff
@@ -0,0 +1,103 @@
+use {
+    crate::{
+        check_program_account,
+        error::TokenError,
+        extension::{
+            group_member_pointer::{
+                instruction::{
+                    GroupMemberPointerInstruction, InitializeInstructionData, UpdateInstructionData,
+                },
+                GroupMemberPointer,
+            },
+            StateWithExtensionsMut,
+        },
+        instruction::{decode_instruction_data, decode_instruction_type},
+        processor::Processor,
+        state::Mint,
+    },
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        entrypoint::ProgramResult,
+        msg,
+        pubkey::Pubkey,
+    },
+    spl_pod::optional_keys::OptionalNonZeroPubkey,
+};
+
+fn process_initialize(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    authority: &OptionalNonZeroPubkey,
+    member_address: &OptionalNonZeroPubkey,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let mint_account_info = next_account_info(account_info_iter)?;
+    let mut mint_data = mint_account_info.data.borrow_mut();
+    let mut mint = StateWithExtensionsMut::<Mint>::unpack_uninitialized(&mut mint_data)?;
+
+    if Option::<Pubkey>::from(*authority).is_none()
+        && Option::<Pubkey>::from(*member_address).is_none()
+    {
+        msg!(
+            "The group member pointer extension requires at least an authority or an address for \
+            initialization, neither was provided"
+        );
+        Err(TokenError::InvalidInstruction)?;
+    }
+
+    let extension = mint.init_extension::<GroupMemberPointer>(true)?;
+    extension.authority = *authority;
+    extension.member_address = *member_address;
+    Ok(())
+}
+
+fn process_update(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    new_member_address: &OptionalNonZeroPubkey,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let mint_account_info = next_account_info(account_info_iter)?;
+    let owner_info = next_account_info(account_info_iter)?;
+    let owner_info_data_len = owner_info.data_len();
+
+    let mut mint_data = mint_account_info.data.borrow_mut();
+    let mut mint = StateWithExtensionsMut::<Mint>::unpack(&mut mint_data)?;
+    let extension = mint.get_extension_mut::<GroupMemberPointer>()?;
+    let authority =
+        Option::<Pubkey>::from(extension.authority).ok_or(TokenError::NoAuthorityExists)?;
+
+    Processor::validate_owner(
+        program_id,
+        &authority,
+        owner_info,
+        owner_info_data_len,
+        account_info_iter.as_slice(),
+    )?;
+
+    extension.member_address = *new_member_address;
+    Ok(())
+}
+
+pub(crate) fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    check_program_account(program_id)?;
```

Comment:
I'm just noticing now that we do this in a few places, but I have no idea why -- we're checking that the executing program is token-2022, but that seems unnecessary. And it makes it annoying for other people to deploy their own version of the program. We can change in a separate PR, but it seems strange.

Reply:
It appears to be coupled to Token-2022 based on the variable name and comment, but technically it's only checking if the provided program ID matches the one generated by `declare_id!(..)`.

https://github.com/solana-labs/solana-program-library/blob/d6bf8540316aff8540700ba3a7f202a55f49cb4c/token/program-2022/src/lib.rs#L100-L106

So, this should work for someone who deploys their own version, since they'll want to change that anyway.

Reply:
It depends. If someone just wants to build the program from source and immediately deploy it, it won't work for certain instructions, which is really weird. Anyway, I put in https://github.com/solana-labs/solana-program-library/issues/5909 for it